src/buchheim.jl

```diff
@@ -1,18 +1,32 @@
+export Buchheim
+
 """
+    Buchheim(; kwargs...)(adj_matrix)
+    Buchheim(; kwargs...)(adj_list)
+    layout(algo::Buchheim, adj_matrix)
+    layout(algo::Buchheim, adj_list)
+
 Using the algorithm proposed in the paper,
-"Improving Walker's Algorithm to Run in Linear Time"
+["Improving Walker's Algorithm to Run in Linear Time"](http://dirk.jivas.de/papers/buchheim02improving.pdf)
 by Christoph Buchheim, Michael Junger, Sebastian Leipert
-(http://dirk.jivas.de/papers/buchheim02improving.pdf)
 
-Arguments
-tree    Adjacency List that represents the given tree
+Takes adjacency matrix or list representation of given tree
+and returns coordinates of the nodes.
+
+## Keyword Arguments
+- `Ptype=Float64`: Determines the output type `Point{2,Ptype}`.
+- `nodesize=Float64[]`
 
-Returns
-positions     co-ordinates of the layout
+  Determines the size of each of the node. If network size does not match the
+  length of `nodesize` fill up with `ones` or truncate given parameter.
 """
-module Buchheim
+struct Buchheim{Ptype,T} <: AbstractLayout{2,Ptype}
+    nodesize::Vector{T}
+end
```

Comment:
When we try to separate the layout parameters from the actually layout calculation, does it make sense to store the node/vertex parameters in this struct? This would mean, that we already need to know how many vertices we have, when we create this struct. Maybe it would be better to specify it in

```julia
layout(::Buchheim; nodsize=nothing) = ...
```

This way, it might also be easier to deal with graphs that have metadata attached to their nodes later.

Reply:
Yes, this is for sure the biggest problem with the proposed interface: it is not completely consistent. There are parameters like the stiffness in spring model or random seeds which are independent of the the the actual layout call. There are other parameters which are tied to stuff like the size of the graph.

I've considered multiple options for this problem:
- `Layout.layout(g; kwargs...)`: the old way. This feels very unjulian to me because it defines several functions rather than methods for the same function. Pros: no additional types needed.
- `layout(::Layout, g; kwargs...)` where `Layout` only stores general parameters. Everything else gets passed to the `layout` function. I think this is quite confusing because there are several points where to inject parameters. In case of graph plotting on need to either define a closure or add some other container to pass the parameters
```julia
lay = g -> layout(Spring(some_para...), g, other_para)
# or
lay = Spring(some_para)
graphplot(g, layout=lay, layoutpara=(other_para...))
```
- `layout(::Layout, g)`: all parameters stored in `Layout` object

I went with the third option. Most of the time it does not matter because users will know the size of their graph and this won't change. Wherever possible I've implemented the restrictions very soft: i.e. if there are more nodes than `nodesize`: just fill up with ones. If there are `N` nodes but `M>N` initial positions? Just take the first `N` etc.

I think this design actually plays very well with stored meta data in case of graph plotting
```julia
g = some_meta_graph()
function mylayout(g::MetaGraph)
    nodesize = extract_data(g)
    return Buchheim(; nodesize)(g)
end
plot(g; layout=mylayout)
```

Reply:
After a short discussion with SimonD i've removed the `layout` function from the public interface. For one-shot application there are now lower case function calls for all of the layouts. So now there are two main ways of applying layouts:
```julia
layoutfunction = MyLayout(; kwargs) # creates callable object to apply layout later
positions = mylayout(g; kwargs...) # apply layout right now
```
the second one is quite similar to the old `Spring.layout(g; kwargs...)`.